src/material/datepicker/calendar.ts

```diff
@@ -149,8 +151,40 @@ export class MatCalendarHeader<D> {
       return this._dateAdapter.getYear(date1) == this._dateAdapter.getYear(date2);
     }
     // Otherwise we are in 'multi-year' view.
-    return Math.floor(this._dateAdapter.getYear(date1) / yearsPerPage) ==
-        Math.floor(this._dateAdapter.getYear(date2) / yearsPerPage);
+    return this._isSameMultiYearView(date1, date2);
+  }
+
+  private _isSameMultiYearView(date1: D, date2: D): boolean {
+    const year1 = this._dateAdapter.getYear(date1);
+    const year2 = this._dateAdapter.getYear(date2);
+    const startingYear = this._getStartingYear();
+    return Math.floor((year1 - startingYear) / yearsPerPage) ===
+            Math.floor((year2 - startingYear) / yearsPerPage);
+  }
+
+  /**
+   * When the multi-year view is first opened, the active year will be in view.
+   * So we compute how many years are between the active year and the *slot* where our
+   * "startingYear" will render when paged into view.
+   */
+  private _getActiveOffset(): number {
+    const activeYear = this._dateAdapter.getYear(this.calendar.activeDate);
+    return euclideanModulo((activeYear - this._getStartingYear()), yearsPerPage);
+  }
+
+  /**
+   * We pick a "starting" year such that either the maximum year would be at the end
+   * or the minimum year would be at the beginning of a page.
+   */
+  private _getStartingYear(): number {
+    let startingYear = 0;
+    if (this.calendar.maxDate) {
+      const maxYear = this._dateAdapter.getYear(this.calendar.maxDate);
+      startingYear = maxYear - yearsPerPage + 1;
+    } else if (this.calendar.minDate) {
+      startingYear = this._dateAdapter.getYear(this.calendar.minDate);
+    }
+    return startingYear;
   }
 }
 
```

Comment:
The 3 methods (getStartingYear, getActiveOffset, and isSameMultiYearView) are now private methods as all of them need the date adapter. Thus, they are now defined in both MatCalendarHeader and MatMultiYearView.